File: tests/07_phases_ompss.dg/fortran/success_task_21.f90
! <testinfo>
! test_generator=config/mercurium-ompss
! </testinfo>
PROGRAM P
    IMPLICIT NONE

    TYPE T0
        INTEGER :: X(10)
    END TYPE T0

    TYPE T2
        INTEGER :: X(10)
    END TYPE T2

    TYPE T1
        TYPE(T2) :: Y
    END TYPE T1

    TYPE(T0) :: VAR1
    TYPE(T1) :: VAR2

    VAR1 % X = -1
    VAR2 % Y % X = -1

    !$OMP TASK INOUT(VAR1 % X)
    VAR1 % X(1) = 1
    !$OMP END TASK
    !$OMP TASKWAIT
    
    IF (VAR1 % X(1) /= 1) STOP 1

    !$OMP TASK INOUT(VAR2 % Y % X)
    VAR2 % Y % X(1) = 1
    !$OMP END TASK
    !$OMP TASKWAIT
    IF (VAR2% Y % X(1) /= 1) STOP 1

END PROGRAM P

File: tests/07_phases_ompss.dg/cxx/success_task_21.cpp
/*
<testinfo>
test_generator=config/mercurium-ompss
</testinfo>
*/
#include <stdlib.h>

template <typename T>
struct A
{
    T t;
    A() : t(0) { }
};

template <typename T>
void f(A<T> *a, int first, int last, int val)
{
#pragma omp task out(a[first:last])
    {
        int i;
        for (i = first; i <= last; i++)
        {
            a[i].t = val;
        }
    }
}

template <typename T>
void g(A<T> *a, int length, int val)
{
#pragma omp task in(a[0;length])
    {
        int i;
        for (i = 0; i < length; i++)
        {
            if (a[i].t != val)
            {
                abort();
            }
        }
    }
}

int main(int argc, char *argv[])
{
    A<int> a[10];

    for (int j = 0; j < 1000; j++)
    {
        f(a, 0, 9, j);
        g(a, 10, j);
    }

#pragma omp taskwait
}
